Flex learner

rbracco opened the pull request:
Alter the learner to accept any number of channels. Most of the code is taken with permission from [Thomas Brandon](https://github.com/thomasbrandon/). Thank you for the help and advice Thomas.

rbracco committed:
Copy code from TomB's repo

rbracco committed:
Modify audio_learner to use correct num channels

rbracco committed:
Add imports from torch.nn

rbracco committed:
remove expansion to 3 channels

rbracco committed:
Merge branch 'master' into flexLearner

Started this branch off an old version of master by accident.

rbracco committed:
fix spectro dims for flexible channels

rbracco committed:
fix spectro display

rbracco committed:
Pass accuracy to learner as default metric

rbracco committed:
fix delta for new learner

rbracco committed:
Alter learning rates for new arch

rbracco committed:
Alter learning rates for new arch

rbracco committed:
Add acknowledgement to Thomas Brandon

Thanks to Thomas Brandon, github.com/thomasbrandon/,
forums.fast.ai/u/tomb/summary for contributing the audio learner code

kevinbird15 commented on audio/data.py:
```diff
@@ -299,10 +299,9 @@ def open(self, item) -> AudioItem:
                 mel = MelSpectrogram(**(cfg.sg_cfg.mel_args()))(sig)
                 if cfg.sg_cfg.to_db_scale: mel = SpectrogramToDB(top_db=cfg.sg_cfg.top_db)(mel)
-            mel = mel.squeeze().permute(1, 0).flip(0)
+            mel = mel.permute(0, 2, 1).flip(1)
```
I think this should be 0,1,2 which would mean we could just drop permute here.  

On my testing, this is how it was at least.

rbracco replied:
Recleared cache, sg is still horizontal. I may have changed stuff elsewhere that affected it and made it differ from yours. But checking this closely did help me to catch and remove `.flip()` as somehow that is causing spectros to be upside down again. 

The only other reason I can think of why youd be experiencing that is the underlying implementation of `MelSpectrogram` in torchaudio may have changed

rbracco replied:
The only reason I can think of why youd be experiencing that is the underlying implementation of `MelSpectrogram` in torchaudio may have changed

kevinbird15 reviewed (commented):
Can you delete your cache and ensure that your spectrograms are horizontal?  For me, I had to permute 0,1,2 instead of 0,2,1 to get it to be horizontal.

rbracco committed:
Skip tfm_pad_signal for now

tfm_pad_signal cant currently be called from within audio.py due
to a cyclic. This causes max_to_pad to not work, so for now we omit
the call and max_to_pad will work but will display the original signal
instead of the padded one that the sg is generated from

rbracco committed:
Remove .flip, to have correct sg orientation

mogwai commented on audio/audio.py:
```diff
@@ -82,7 +79,8 @@ def shape(self): return self.data.shape
         if self.max_to_pad is not None:
-            sig = tfm_padtrim_signal(sig, int(self.max_to_pad/1000*sr), pad_mode="zeros")
+            pass
+            # sig = tfm_padtrim_signal(sig, int(self.max_to_pad/1000*sr), pad_mode="zeros")
```
remove comments ?

mogwai commented on audio/learner.py:
```diff
@@ -1,9 +1,61 @@
+# This will have to be updated in future for multichannel but is a quick fix for now
+def _calc_channels(cfg):
+    channels = 3 if cfg.delta else 1
+    return channels
```
Can we have a credit comment here?

rbracco replied:
There is a credit comment above the audio learner, and in the commit  message. I actually wrote this particular function. Lmk if you want additional, maybe there should be one on the PR in general.

mogwai replied:
Yeah ok sorry i thought you would credit each function but thats fine

mogwai replied:
Should have been a seperate PR yes

rbracco replied:
Adding only his code breaks things so it can't be its own PR. We had to alter lots of stuff elsewhere and write the `_calc_channels` helper func to make it work, that's what the rest of the commits are.

rbracco committed:
remove unnecessary comments
